### ISSUE TITLE
Ruby 3.1 and Ruby 3.2 (with YJIT) benchmarks

### ISSUE THREAD

**Issue by Willianvdv** [rmosolgo/graphql-ruby#4281]
With all the talk about the performance gains with Ruby 3.2 and its YJIT compiler, I was kinda curious what the effect on GraphQL-Ruby would be. I ran the parse, query, and validate benchmarks for Ruby 3.1 and Ruby 3.2 (with YJIT enabled). I'm using an Apple M1 Pro and these benchmarks run in docker containers (based on https://hub.docker.com/_/ruby/).

# Benchmark - Parse

## Ruby 3.1
```bash
root@b5bceda237f1:/app# ruby -v ; bundle exec rake bench:parse
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [aarch64-linux]
Warming up --------------------------------------
scan - introspection   118.000  i/100ms
parse - introspection
                        94.000  i/100ms
    scan - fragments   211.000  i/100ms
   parse - fragments   168.000  i/100ms
    scan - big query    20.000  i/100ms
   parse - big query    16.000  i/100ms
Calculating -------------------------------------
scan - introspection      1.173k (± 2.7%) i/s -      5.900k in   5.032889s
parse - introspection
                        953.735  (± 1.5%) i/s -      4.794k in   5.027665s
    scan - fragments      2.103k (± 2.8%) i/s -     10.550k in   5.021119s
   parse - fragments      1.657k (± 2.5%) i/s -      8.400k in   5.073460s
    scan - big query    208.255  (± 1.4%) i/s -      1.060k in   5.090883s
   parse - big query    167.591  (± 1.2%) i/s -    848.000  in   5.060728s
```

## Ruby 3.2 with YJIT
```bash
root@12f6577c7ff1:/app# export RUBY_YJIT_ENABLE=1 ; ruby -v ; bundle exec rake bench:parse
ruby 3.2.0rc1 (2022-12-23) +YJIT [aarch64-linux]
Warming up --------------------------------------
scan - introspection   147.000  i/100ms
parse - introspection
                       113.000  i/100ms
    scan - fragments   252.000  i/100ms
   parse - fragments   194.000  i/100ms
    scan - big query    25.000  i/100ms
   parse - big query    19.000  i/100ms
Calculating -------------------------------------
scan - introspection      1.467k (± 1.6%) i/s -      7.350k in   5.012185s
parse - introspection
                          1.147k (± 1.3%) i/s -      5.763k in   5.024536s
    scan - fragments      2.497k (± 2.2%) i/s -     12.600k in   5.048754s
   parse - fragments      1.937k (± 1.5%) i/s -      9.700k in   5.009974s
    scan - big query    248.782  (± 1.2%) i/s -      1.250k in   5.025143s
   parse - big query    193.295  (± 1.6%) i/s -    969.000  in   5.014247s
```

# Benchmark - Query

## Ruby 3.1
```bash
root@b5bceda237f1:/app# ruby -v ; bundle exec rake bench:query
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [aarch64-linux]
Warming up --------------------------------------
               query     3.000  i/100ms
Calculating -------------------------------------
               query     31.540  (± 3.2%) i/s -    159.000  in   5.044089s
```

## Ruby 3.2 with YJIT
```bash
root@12f6577c7ff1:/app# export RUBY_YJIT_ENABLE=1 ; ruby -v ; bundle exec rake bench:query
ruby 3.2.0rc1 (2022-12-23) +YJIT [aarch64-linux]
Warming up --------------------------------------
               query     4.000  i/100ms
Calculating -------------------------------------
               query     43.262  (± 6.9%) i/s -    216.000  in   5.016860s
```

# Benchmark - Validate
## Ruby 3.1
```bash
root@b5bceda237f1:/app# ruby -v ; bundle exec rake bench:validate
ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [aarch64-linux]
Warming up --------------------------------------
validate - introspection
                       144.000  i/100ms
validate - abstract fragments
                       270.000  i/100ms
validate - abstract fragments 2
                       192.000  i/100ms
validate - big query    15.000  i/100ms
validate - fields will merge
                         1.000  i/100ms
Calculating -------------------------------------
validate - introspection
                          1.484k (± 2.4%) i/s -      7.488k in   5.049395s
validate - abstract fragments
                          2.668k (± 2.0%) i/s -     13.500k in   5.061057s
validate - abstract fragments 2
                          1.914k (± 1.7%) i/s -      9.600k in   5.018288s
validate - big query    157.114  (± 3.8%) i/s -    795.000  in   5.067433s
validate - fields will merge
                          0.223  (± 0.0%) i/s -      2.000  in   8.979272s
```

## Ruby 3.2 with YJIT
```bash
root@12f6577c7ff1:/app# export RUBY_YJIT_ENABLE=1 ; ruby -v ; bundle exec rake bench:validate
ruby 3.2.0rc1 (2022-12-23) +YJIT [aarch64-linux]
Warming up --------------------------------------
validate - introspection
                       203.000  i/100ms
validate - abstract fragments
                       387.000  i/100ms
validate - abstract fragments 2
                       272.000  i/100ms
validate - big query    20.000  i/100ms
validate - fields will merge
                         1.000  i/100ms
Calculating -------------------------------------
validate - introspection
                          2.157k (± 1.7%) i/s -     10.962k in   5.082739s
validate - abstract fragments
                          3.975k (± 2.1%) i/s -     20.124k in   5.065130s
validate - abstract fragments 2
                          2.688k (± 3.2%) i/s -     13.600k in   5.065149s
validate - big query    207.967  (± 3.8%) i/s -      1.040k in   5.009594s
validate - fields will merge
                          0.727  (± 0.0%) i/s -      4.000  in   5.500702s
```

**Comment by rmosolgo** [comment]
Hi again! Long time no see 😁 , happy new year. 

In short, it looks like the takeaways is that GraphQL-Ruby _should_ benefit from Ruby 3.2 + YJIT, isn't that right? 

Besides this, I'm hoping to start working closely with Shopify this year to improve runtime performance specifically, and part of that work might be making GraphQL-Ruby more ... YJIT-able. So there may be more to come!

**Comment by Willianvdv** [comment]
> Hi again! Long time no see 😁 , happy new year.

You too! 🎆 

> In short, it looks like the takeaways is that GraphQL-Ruby _should_ benefit from Ruby 3.2 + YJIT, isn't that right?

Yeah, looking at the benchmarks there's a noticeable improvement! I was hoping YJIT would help complex GraphQL calls where we visit the same methods over and over. To be honest, I'm not sure at this moment how much of this is due to YJIT or one of the other optimizations shipped in Ruby 3.2. In any case, these results make me very excited to upgrade our app to 3.2 and see what the effects are on our GraphQL endpoint on production.

> Besides this, I'm hoping to start working closely with Shopify this year to improve runtime performance specifically, and part of that work might be making GraphQL-Ruby more ... YJIT-able. So there may be more to come!

Very cool! What kind of things do you think would help the YJIT compiler?
